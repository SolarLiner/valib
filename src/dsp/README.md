Base traits for DSP algorithms.

Includes 3 main traits that algorithms can implement:

- the [`DSP`] trait defines the main methods to implement for per-sample algorithms.
  Most algorithms implements this version.
- the [`DSPBlock`] trait defines the methods to implement for per-block processing.
  Also includes helpers for using block processing in a per-sample context.
  [`DSP`] implementers get an impl for this for free. This means that
  algorhithms that would benefit from a specialized per-block processing impl
  cannot currently do it. There is a fix coming soon.
- the [`DspAnalysis`] trait defines the methods to provide frequency and phase
  response analysis.