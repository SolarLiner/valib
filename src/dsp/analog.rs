use nalgebra::Complex;
use crate::dsp::DSP;

pub trait DspAnalog<const I: usize, const O: usize>: DSP<I, O> {
    fn h_s(&self, s: [Complex<Self::Sample>; I]) -> [Complex<Self::Sample>; O];
}
