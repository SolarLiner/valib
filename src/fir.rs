use std::{collections::VecDeque, ops};

use numeric_literals::replace_float_literals;

use crate::dsp::DSPMeta;
use crate::{dsp::DSPProcess, Scalar};

fn slice_add<T: Copy + ops::Add<T, Output = T>>(in1: &[T], in2: &[T], out: &mut [T]) {
    let len = in1.len().min(in2.len()).min(out.len());
    for i in 0..len {
        out[i] = in1[i] + in2[i];
    }
}

fn slice_add_to<T: Copy + ops::AddAssign<T>>(out: &mut [T], inp: &[T]) {
    for (o, i) in out.iter_mut().zip(inp.iter().copied()) {
        *o += i;
    }
}

fn slice_sub_to<T: Copy + ops::SubAssign<T>>(out: &mut [T], inp: &[T]) {
    for (o, i) in out.iter_mut().zip(inp.iter().copied()) {
        *o -= i;
    }
}

/// Perform arbitrary-length convolution with an O(n^log2(3)) complexity
/// Translated from Taken from https://www.musicdsp.org/en/latest/Filters/65-time-domain-convolution-with-o-n-log2-3.html
/// in1, in2: input signals (of size n)
/// out: output signal (of size 2n)
/// buffer: staging buffer (of size 2n)
#[replace_float_literals(T::from_f64(literal))]
#[allow(unused)]
fn convolution<T: Scalar>(in1: &[T], in2: &[T], out: &mut [T], buffer: &mut [T]) {
    let size = in1.len();
    debug_assert_eq!(size, in2.len());
    debug_assert_eq!(out.len(), 2 * size);
    debug_assert_eq!(buffer.len(), 2 * size);

    out.fill(0.0);

    if size == 1 {
        out[0] = in1[0] * in2[0];
        return;
    }

    // first calculate (A1 + A2 * z^-n)*(B1 + B2 * z^-n)*z^n
    {
        let (in11, in12) = in1.split_at(size / 2);
        slice_add(in11, in12, buffer);
    }
    {
        let (in21, in22) = in2.split_at(size / 2);
        slice_add(in21, in22, &mut buffer[size / 2..]);
    }
    {
        let (tmp1, tmp2) = buffer.split_at_mut(size / 2);
        let (tmp2, tmp3) = tmp2.split_at_mut(size);
        convolution(tmp1, tmp2, &mut out[size / 2..], tmp3);
    }

    // then add A1*B1 and substract A1*B1*z^n
    {
        let (tmp1, tmp2) = buffer.split_at_mut(size);
        convolution(
            &in1[..size / 2],
            &in2[..size / 2],
            &mut tmp1[..size / 2],
            &mut tmp2[..size / 2],
        );
        slice_add_to(out, tmp1);
        slice_sub_to(&mut out[size / 2..], &tmp1[..size / 2]);
    }

    // then add A2*B2 and substract A2*B2*z^-n
    {
        let (tmp1, tmp2) = buffer.split_at_mut(size);
        convolution(
            &in1[size / 2..],
            &in2[size / 2..],
            &mut tmp1[..size / 2],
            &mut tmp2[..size / 2],
        );
        slice_add_to(&mut out[size..], tmp1);
        slice_sub_to(&mut out[size / 2..], tmp1);
    }
}

pub struct Fir<T> {
    kernel: Box<[T]>,
    memory: VecDeque<T>,
    kernel_latency: usize,
}

impl<T: Scalar> Fir<T> {
    pub fn lowpass(fc: T, bandwidth: f64) -> Self {
        let kernel = Vec::from(kernels::windowed_sinc(fc, bandwidth));
        let len = kernel.len();
        Self::new(kernel, len / 2)
    }

    pub fn new(kernel: impl IntoIterator<Item = T>, kernel_latency: usize) -> Self {
        let kernel = Box::from_iter(kernel);
        let memory = VecDeque::from(vec![T::from_f64(0.0); kernel.len()]);
        Self {
            kernel,
            memory,
            kernel_latency,
        }
    }
}

impl<T: Scalar> DSPMeta for Fir<T> {
    type Sample = T;

    fn latency(&self) -> usize {
        self.kernel_latency
    }

    fn reset(&mut self) {
        self.memory = VecDeque::from(vec![T::from_f64(0.0); self.kernel.len()]);
    }
}

impl<T: Scalar> DSPProcess<1, 1> for Fir<T> {
    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let [x] = x;
        self.memory.pop_front();
        self.memory.push_back(x);
        let y = self
            .kernel
            .iter()
            .copied()
            .zip(self.memory.iter().copied())
            .fold(T::from_f64(0.0), |acc, (a, b)| acc + a * b);
        [y]
    }
}

pub mod kernels {
    use numeric_literals::replace_float_literals;

    use crate::Scalar;

    #[replace_float_literals(T::from_f64(literal))]
    pub fn windowed_sinc_in_place<T: Scalar>(fc: T, slice: &mut [T]) {
        debug_assert_eq!(slice.len() % 2, 1);
        let width = T::from_f64(slice.len() as _);
        let half_width = 0.5 * width;
        for (i, s) in slice.iter_mut().enumerate() {
            let i = T::from_f64(i as _);
            *s = T::simd_sin(T::simd_two_pi() * fc * (i - half_width)) / (i - half_width);
            *s *= 0.42 - 0.5 * T::simd_cos(T::simd_two_pi() * i / width)
                + 0.08 * T::simd_cos(2.0 * T::simd_two_pi() * i / width);
        }
        // Normalization
        let magnitude = slice.iter().copied().fold(0.0, |a, b| a + b);
        for s in slice {
            *s /= magnitude;
        }
    }

    pub fn windowed_sinc<T: Scalar>(fc: T, bandwidth: f64) -> Box<[T]> {
        let mut length = (4.0 / bandwidth) as usize;
        if length % 2 == 0 {
            length += 1;
        }

        let mut data = vec![T::from_f64(0.0); length];
        windowed_sinc_in_place(fc, &mut data[..]);
        data.into_boxed_slice()
    }
}

#[cfg(test)]
mod tests {
    use crate::dsp::buffer::AudioBuffer;
    use crate::dsp::DSPProcessBlock;

    use super::*;

    #[test]
    fn test_fir_direct() {
        let input = Box::from_iter([1.0, 0.0, 0.0, 0.0].into_iter().cycle().take(16));
        let input = AudioBuffer::new([input]).unwrap();
        let mut output = input.clone();
        let mut fir = Fir::new([0.25, 0.5, 0.25], 1);

        output.fill(0.0);
        fir.process_block(input.as_ref(), output.as_mut());
        insta::assert_csv_snapshot!(output.get_channel(0), { "[]" => insta::rounded_redaction(4) })
    }

    #[test]
    fn test_kernel_lowpass() {
        insta::assert_csv_snapshot!(&kernels::windowed_sinc(0.1, 0.1), { "[]" => insta::rounded_redaction(5) });
    }
}
