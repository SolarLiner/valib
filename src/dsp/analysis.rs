use nalgebra::{Complex, zero};
use simba::simd::SimdComplexField;
use numeric_literals::replace_float_literals;
use simba::scalar::RealField;
use crate::dsp::DSP;
use crate::Scalar;
use crate::dsp::analog::DspAnalog;

pub trait DspAnalysis<const I: usize, const O: usize>: DSP<I, O> {
    fn h_z(&self, z: [Complex<Self::Sample>; I]) -> [Complex<Self::Sample>; O];
    fn freq_response(&self, jw: [Self::Sample; I]) -> [Complex<Self::Sample>; O]
        where
            Self::Sample: RealField,
            Complex<Self::Sample>: SimdComplexField<SimdRealField=Self::Sample>
    {
        let z = jw.map(Complex::from_simd_real).map(|jw| <Complex<Self::Sample> as SimdComplexField>::simd_exp(Complex::i() * jw));
        self.h_z(z)
    }
}

impl<const I: usize, const O: usize, D: DspAnalog<I, O>> DspAnalysis<I, O> for D where Complex<D::Sample>: SimdComplexField {
    #[replace_float_literals(Complex::from(D::Sample::from_f64(literal)))]
    #[inline(always)]
    fn h_z(&self, z: [Complex<D::Sample>; I]) -> [Complex<D::Sample>; O] {
        self.h_s(z.map(|z| 2. * (z - 1.) / (z + 1.)))
    }

    #[inline(always)]
    fn freq_response(&self, jw: [D::Sample; I]) -> [Complex<D::Sample>; O]
        where
            Self::Sample: RealField,
    {
        self.h_s(jw.map(|jw| Complex::new(D::Sample::from_f64(0.0), jw)))
    }
}
