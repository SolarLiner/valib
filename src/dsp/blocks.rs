//! Small [`DSP`] building blocks for reusability.
use std::marker::PhantomData;
use std::ops;
use std::ops::DerefMut;

use nalgebra::{Complex, ComplexField, SMatrix, SVector};
use num_traits::{One, Zero};
use numeric_literals::replace_float_literals;

use crate::dsp::analysis::DspAnalysis;
use crate::dsp::DSP;
use crate::Scalar;

/// "Bypass" struct, which simply forwards the input to the output.
#[derive(Debug, Copy, Clone, Default)]
pub struct Bypass<S>(PhantomData<S>);

impl<S: Scalar, const N: usize> DSP<N, N> for Bypass<S> {
    type Sample = S;

    fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
        x
    }
}

/// Freestanding integrator, discretized with TPT
#[derive(Debug, Copy, Clone)]
pub struct Integrator<T>(pub T);

impl<T: Scalar> Default for Integrator<T> {
    fn default() -> Self {
        Self(T::zero())
    }
}

impl<T: Scalar> DSP<1, 1> for Integrator<T> {
    type Sample = T;

    #[replace_float_literals(T::from_f64(literal))]
    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 1] {
        let x2 = x[0] / 2.0;
        let out = x2 + self.0;
        self.0 = x2 + out;
        [out]
    }
}

impl<T: Scalar> DspAnalysis<1, 1> for Integrator<T> {
    #[replace_float_literals(Complex::from(T::from_f64(literal)))]
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; 1]; 1] {
        [[1. / 2. * (z + 1.) / (z - 1.)]]
    }
}

/// Sum inputs into a single output
#[derive(Debug, Copy, Clone, Default)]
pub struct Sum<T, const N: usize>(PhantomData<[T; N]>);

impl<T, const N: usize> DSP<N, 1> for Sum<T, N>
where
    T: Scalar,
{
    type Sample = T;

    fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; 1] {
        [x.into_iter().fold(T::zero(), |a, b| a + b)]
    }
}

impl<T, const N: usize> DspAnalysis<N, 1> for Sum<T, N>
where
    Self: DSP<N, 1>,
{
    fn h_z(&self, _z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; 1]; N] {
        [[Complex::one()]; N]
    }
}

/// 6 dB/oct one-pole filter using the "one-sample trick" (fig. 3.31, eq. 3.32).
/// Outputs modes as follows: [LP, HP, AP].
#[derive(Debug, Copy, Clone)]
pub struct P1<T> {
    w_step: T,
    fc: T,
    s: T,
}

impl<T: Scalar> DspAnalysis<1, 3> for P1<T>
where
    Self::Sample: nalgebra::RealField,
{
    #[replace_float_literals(Complex::from_real(< T as Scalar >::from_f64(literal)))]
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; 3]; 1] {
        let lp = (z - 1.0) / (z + 1.0) * self.fc / 2.0;
        let hp = 1.0 - lp;
        let ap = 2.0 * lp - 1.0;
        [[lp, hp, ap]]
    }
}

impl<T: Scalar> P1<T> {
    pub fn new(samplerate: T, fc: T) -> Self {
        Self {
            w_step: T::simd_pi() / samplerate,
            fc,
            s: T::zero(),
        }
    }

    pub fn with_state(mut self, state: T) -> Self {
        self.s = state;
        self
    }

    pub fn set_fc(&mut self, fc: T) {
        self.fc = fc;
    }
}

impl<T: Scalar> DSP<1, 3> for P1<T> {
    type Sample = T;

    #[inline(always)]
    #[replace_float_literals(T::from_f64(literal))]
    fn process(&mut self, x: [Self::Sample; 1]) -> [Self::Sample; 3] {
        // One-sample feedback trick over a transposed integrator, implementation following
        // eq (3.32), page 77
        let g = self.w_step * self.fc;
        let k = g / (1. + g);
        let v = k * (x[0] - self.s);
        let lp = v + self.s;
        self.s = lp + v;

        let hp = x[0] - lp;
        let ap = 2. * lp - x[0];
        [lp, hp, ap]
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        self.w_step = T::simd_pi() / T::from_f64(samplerate as _)
    }

    fn latency(&self) -> usize {
        1
    }

    fn reset(&mut self) {
        self.s = T::zero();
    }
}

/// Process inner DSP blocks in series. `DSP` is implemented for tuples up to 8 elements all of the same I/O configuration.
#[derive(Debug, Copy, Clone)]
pub struct Series<T>(pub T);

macro_rules! series_tuple {
    ($($p:ident),*) => {
        #[allow(non_snake_case)]
        impl<__Sample: $crate::Scalar, $($p: $crate::dsp::DSP<N, N, Sample = __Sample>),*, const N: usize> DSP<N, N> for $crate::dsp::blocks::Series<($($p),*)> {
            type Sample = __Sample;

            fn latency(&self) -> usize {
                let Self(($($p),*)) = self;
                0 $(
                + $p.latency()
                )*
            }

            fn reset(&mut self) {
                let Self(($($p),*)) = self;
                $(
                $p.reset();
                )*
            }

            #[allow(non_snake_case)]
            #[inline(always)]
            fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
                let Self(($($p),*)) = self;
                $(
                let x = $p.process(x);
                )*
                x
            }

            fn set_samplerate(&mut self, samplerate: f32) {
                let Self(($($p),*)) = self;
                $(
                $p.set_samplerate(samplerate);
                )*
            }
        }
    };
}

series_tuple!(A, B);
series_tuple!(A, B, C);
series_tuple!(A, B, C, D);
series_tuple!(A, B, C, D, E);
series_tuple!(A, B, C, D, E, F);
series_tuple!(A, B, C, D, E, F, G);
series_tuple!(A, B, C, D, E, F, G, H);

/// Specialized `Series` struct that doesn't restrict the I/O count of either DSP struct
#[derive(Debug, Copy, Clone)]
pub struct Series2<A, B, const INNER: usize>(A, PhantomData<[(); INNER]>, B);

impl<A, B, const INNER: usize> Series2<A, B, INNER> {
    /// Construct a new `Series2` instance, with each inner DSP instance given.
    pub const fn new<const I: usize, const O: usize>(a: A, b: B) -> Self
    where
        A: DSP<I, INNER>,
        B: DSP<INNER, O>,
    {
        Self(a, PhantomData, b)
    }

    /// Returns a reference to the first DSP instance, which processes the incoming audio first.
    pub const fn left(&self) -> &A {
        &self.0
    }

    /// Returns a mutable reference to the first DSP instance, which processes the incoming audio first.
    pub fn left_mut(&mut self) -> &mut A {
        &mut self.0
    }

    /// Returns a reference to the second DSP instance, which processes the incoming audio last.
    pub const fn right(&self) -> &B {
        &self.2
    }

    /// Returns a mutable reference to the second DSP instance, which processes the incoming audio last.
    pub fn right_mut(&mut self) -> &mut B {
        &mut self.2
    }
}

impl<A, B, const I: usize, const J: usize, const O: usize> DSP<I, O> for Series2<A, B, J>
where
    A: DSP<I, J>,
    B: DSP<J, O, Sample = A::Sample>,
{
    type Sample = A::Sample;

    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        let Self(a, _, b) = self;
        let j = a.process(x);
        b.process(j)
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        self.0.set_samplerate(samplerate);
        self.2.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        let Self(a, _, b) = self;
        a.latency() + b.latency()
    }

    fn reset(&mut self) {
        let Self(a, _, b) = self;
        a.reset();
        b.reset();
    }
}

impl<P: DSP<N, N>, const N: usize, const C: usize> DSP<N, N> for Series<[P; C]> {
    type Sample = P::Sample;

    fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
        self.0.iter_mut().fold(x, |x, dsp| dsp.process(x))
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        for s in &mut self.0 {
            s.set_samplerate(samplerate);
        }
    }

    fn latency(&self) -> usize {
        self.0.iter().map(|dsp| dsp.latency()).sum()
    }

    fn reset(&mut self) {
        for dsp in self.0.iter_mut() {
            dsp.reset();
        }
    }
}

impl<P, const N: usize> DSP<N, N> for Series<Vec<P>>
where
    P: DSP<N, N>,
{
    type Sample = P::Sample;

    fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
        self.0.iter_mut().fold(x, |x, dsp| dsp.process(x))
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        for s in &mut self.0 {
            s.set_samplerate(samplerate);
        }
    }

    fn latency(&self) -> usize {
        self.0.iter().map(|dsp| dsp.latency()).sum()
    }

    fn reset(&mut self) {
        for dsp in self.0.iter_mut() {
            dsp.reset();
        }
    }
}

impl<P, const N: usize, const C: usize> DspAnalysis<N, N> for Series<[P; C]>
where
    P: DspAnalysis<N, N>,
{
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; N]; N] {
        self.0.iter().fold([[Complex::one(); N]; N], |acc, f| {
            let ret = f.h_z(z);
            std::array::from_fn(|i| std::array::from_fn(|j| acc[i][j] * ret[i][j]))
        })
    }
}

impl<P, const N: usize> DspAnalysis<N, N> for Series<Vec<P>>
where
    P: DspAnalysis<N, N>,
{
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; N]; N] {
        self.0.iter().fold([[Complex::one(); N]; N], |acc, f| {
            let ret = f.h_z(z);
            std::array::from_fn(|i| std::array::from_fn(|j| acc[i][j] * ret[i][j]))
        })
    }
}

/// Process inner DSP blocks in parallel. Input is fanned out to all inner blocks, then summed back out.
#[derive(Debug, Copy, Clone)]
pub struct Parallel<T>(pub T);

macro_rules! parallel_tuple {
    ($($p:ident),*) => {
        #[allow(non_snake_case)]
        impl<__Sample: $crate::Scalar, $($p: $crate::dsp::DSP<N, N, Sample = __Sample>),*, const N: usize> $crate::dsp::DSP<N, N> for $crate::dsp::blocks::Parallel<($($p),*)> {
            type Sample = __Sample;

            fn latency(&self) -> usize {
                let Self(($($p),*)) = self;
                let latency = 0;
                $(
                let latency = latency.max($p.latency());
                )*
                latency
            }

            fn reset(&mut self) {
                let Self(($($p),*)) = self;
                $(
                $p.reset();
                )*
            }

            #[allow(non_snake_case)]
            #[inline(always)]
            fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
                let Self(($($p),*)) = self;
                let mut ret = [Self::Sample::zero(); N];
                $(
                let y = $p.process(x);
                for i in 0..N {
                    ret[i] += y[i];
                }
                )*
                ret
            }

            fn set_samplerate(&mut self, samplerate: f32) {
                let Self(($($p),*)) = self;
                $(
                $p.set_samplerate(samplerate);
                )*
            }
        }
    };
}

parallel_tuple!(A, B);
parallel_tuple!(A, B, C);
parallel_tuple!(A, B, C, D);
parallel_tuple!(A, B, C, D, E);
parallel_tuple!(A, B, C, D, E, F);
parallel_tuple!(A, B, C, D, E, F, G);
parallel_tuple!(A, B, C, D, E, F, G, H);

impl<P: DSP<I, O>, const I: usize, const O: usize, const N: usize> DSP<I, O> for Parallel<[P; N]> {
    type Sample = P::Sample;

    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        self.0
            .iter_mut()
            .map(|dsp| dsp.process(x))
            .fold([Self::Sample::from_f64(0.0); O], |out, dsp| {
                std::array::from_fn(|i| out[i] + dsp[i])
            })
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        for s in &mut self.0 {
            s.set_samplerate(samplerate);
        }
    }

    fn latency(&self) -> usize {
        self.0.iter().fold(0, |max, dsp| max.max(dsp.latency()))
    }

    fn reset(&mut self) {
        for dsp in self.0.iter_mut() {
            dsp.reset();
        }
    }
}

impl<P, const I: usize, const O: usize, const N: usize> DspAnalysis<I, O> for Parallel<[P; N]>
where
    P: DspAnalysis<I, O>,
{
    fn h_z(&self, z: Complex<Self::Sample>) -> [[Complex<Self::Sample>; O]; I] {
        self.0.iter().fold([[Complex::zero(); O]; I], |acc, f| {
            let ret = f.h_z(z);
            std::array::from_fn(|i| std::array::from_fn(|j| acc[i][j] + ret[i][j]))
        })
    }
}

/// Mod matrix struct, with direct access to the summing matrix
#[derive(Debug, Copy, Clone)]
pub struct ModMatrix<T, const I: usize, const O: usize> {
    /// Mod matrix weights, setup in column-major form to produce outputs from inputs with a single matrix-vector
    /// multiplication.
    pub weights: SMatrix<T, O, I>,
}

impl<T, const I: usize, const O: usize> Default for ModMatrix<T, I, O>
where
    T: Scalar,
{
    fn default() -> Self {
        Self {
            weights: SMatrix::from([[T::from_f64(0.0); O]; I]),
        }
    }
}

impl<T, const I: usize, const O: usize> DSP<I, O> for ModMatrix<T, I, O>
where
    T: Scalar,
{
    type Sample = T;

    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        let res = self.weights * SVector::from(x);
        std::array::from_fn(|i| res[i])
    }
}

/// Feedback adapter with a one-sample delay and integrated mixing and summing point.
pub struct Feedback<FF, FB, const N: usize>
where
    FF: DSP<N, N>,
{
    memory: [FF::Sample; N],
    /// Inner DSP instance
    pub feedforward: FF,
    pub feedback: FB,
    /// Mixing vector, which is lanewise-multiplied from the output and summed back to the input at the next sample.
    pub mix: [FF::Sample; N],
}

impl<FF: DSP<N, N>, const N: usize> DSP<N, N> for Feedback<FF, (), N> {
    type Sample = FF::Sample;

    fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
        let x = std::array::from_fn(|i| self.memory[i] * self.mix[i] + x[i]);
        let y = self.feedforward.process(x);
        self.memory = y;
        y
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        self.feedforward.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        self.feedforward.latency()
    }

    fn reset(&mut self) {
        self.memory.fill(Self::Sample::zero());
        self.feedforward.reset();
    }
}

impl<FF, FB, const N: usize> DSP<N, N> for Feedback<FF, FB, N>
where
    FF: DSP<N, N>,
    FB: DSP<N, N, Sample = <FF as DSP<N, N>>::Sample>,
{
    type Sample = FF::Sample;

    fn process(&mut self, x: [Self::Sample; N]) -> [Self::Sample; N] {
        let fb = self.feedback.process(self.memory);
        let x = std::array::from_fn(|i| fb[i] * self.mix[i] + x[i]);
        let y = self.feedforward.process(x);
        self.memory = y;
        y
    }

    fn set_samplerate(&mut self, samplerate: f32) {
        self.feedforward.set_samplerate(samplerate);
        self.feedback.set_samplerate(samplerate);
    }

    fn latency(&self) -> usize {
        self.feedforward.latency()
    }

    fn reset(&mut self) {
        self.memory.fill(FB::Sample::from_f64(0.0));
        self.feedforward.reset();
        self.feedback.reset();
    }
}

impl<FF: DSP<N, N>, FB, const N: usize> Feedback<FF, FB, N> {
    /// Create a new Feedback adapter with the provider inner DSP instance. Sets the mix to 0 by default.
    pub fn new(feedforward: FF, feedback: FB) -> Self {
        Self {
            memory: [FF::Sample::from_f64(0.0); N],
            feedforward,
            feedback,
            mix: [FF::Sample::from_f64(0.0); N],
        }
    }

    /// Unwrap this adapter and give back the inner DSP instance.
    pub fn into_inner(self) -> (FF, FB) {
        (self.feedforward, self.feedback)
    }
}
