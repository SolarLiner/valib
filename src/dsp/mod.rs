use crate::Scalar;

use self::analysis::DspAnalysis;

pub mod analog;
pub mod analysis;
pub mod blocks;
pub mod utils;

pub trait DSP<const I: usize, const O: usize> {
    type Sample: Scalar;

    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O];

    fn latency(&self) -> usize {
        0
    }

    fn reset(&mut self) {}
}

pub trait DSPBlock<const I: usize, const O: usize> {
    type Sample: Scalar;

    /// Caller must ensure inputs and outputs buffers are of the same size.
    fn process_block(&mut self, inputs: &[[Self::Sample; I]], outputs: &mut [[Self::Sample; O]]);

    #[inline(always)]
    fn max_block_size(&self) -> Option<usize> {
        None
    }

    #[inline(always)]
    fn latency(&self) -> usize {
        0
    }

    #[inline(always)]
    fn reset(&mut self) {}
}

impl<P, const I: usize, const O: usize> DSPBlock<I, O> for P
where
    P: DSP<I, O>,
{
    type Sample = <Self as DSP<I, O>>::Sample;

    #[inline(never)]
    fn process_block(&mut self, inputs: &[[Self::Sample; I]], outputs: &mut [[Self::Sample; O]]) {
        let len = inputs.len();
        for i in 0..len {
            outputs[i] = self.process(inputs[i]);
        }
    }

    #[inline(always)]
    fn latency(&self) -> usize {
        DSP::latency(self)
    }

    #[inline(always)]
    fn reset(&mut self) {
        DSP::reset(self)
    }
}

pub struct PerSampleBlockAdapter<P, const I: usize, const O: usize>
where
    P: DSPBlock<I, O>,
{
    input_buffer: Box<[[P::Sample; I]]>,
    input_filled: usize,
    output_buffer: Box<[[P::Sample; O]]>,
    output_filled: usize,
    inner: P,
}

impl<P, const I: usize, const O: usize> std::ops::Deref for PerSampleBlockAdapter<P, I, O>
where
    P: DSPBlock<I, O>,
{
    type Target = P;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<P, const I: usize, const O: usize> std::ops::DerefMut for PerSampleBlockAdapter<P, I, O>
where
    P: DSPBlock<I, O>,
{
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

impl<P, const I: usize, const O: usize> PerSampleBlockAdapter<P, I, O>
where
    P: DSPBlock<I, O>,
{
    pub const DEFAULT_BUFFER_SIZE: usize = 64;
    pub fn new(dsp_block: P) -> Self {
        Self::new_with_max_buffer_size(dsp_block, Self::DEFAULT_BUFFER_SIZE)
    }

    pub fn new_with_max_buffer_size(dsp_block: P, max_buffer_size: usize) -> Self {
        let buffer_size = dsp_block
            .max_block_size()
            .map(|mbs| mbs.min(max_buffer_size))
            .unwrap_or(max_buffer_size);
        Self {
            input_buffer: vec![[P::Sample::from_f64(0.0); I]; buffer_size].into_boxed_slice(),
            input_filled: 0,
            output_buffer: vec![[P::Sample::from_f64(0.0); O]; buffer_size].into_boxed_slice(),
            output_filled: buffer_size,
            inner: dsp_block,
        }
    }

    pub fn into_inner(self) -> P {
        self.inner
    }
}

impl<P, const I: usize, const O: usize> DSP<I, O> for PerSampleBlockAdapter<P, I, O>
where
    P: DSPBlock<I, O>,
{
    type Sample = P::Sample;

    fn process(&mut self, x: [Self::Sample; I]) -> [Self::Sample; O] {
        self.input_buffer[self.input_filled] = x;
        self.input_filled += 1;
        if self.input_buffer.len() == self.input_filled {
            self.inner
                .process_block(&self.input_buffer, &mut self.output_buffer);
            self.input_filled = 0;
            self.output_filled = 0;
        }

        if self.output_filled < self.output_buffer.len() {
            let ret = self.output_buffer[self.output_filled];
            self.output_filled += 1;
            ret
        } else {
            [Self::Sample::from_f64(0.0); O]
        }
    }

    fn latency(&self) -> usize {
        (self.inner.latency() + self.input_buffer.len()).saturating_sub(1)
    }

    fn reset(&mut self) {
        self.inner.reset();
        self.input_filled = 0;
        self.output_filled = self.output_buffer.len();
    }
}

impl<P, const I: usize, const O: usize> DspAnalysis<I, O> for PerSampleBlockAdapter<P, I, O>
where P: DspAnalysis<I, O>
{
    fn h_z(&self, z: [nalgebra::Complex<Self::Sample>; I]) -> [nalgebra::Complex<Self::Sample>; O] {
        self.inner.h_z(z)
    }
}


#[cfg(test)]
mod tests {
    use std::marker::PhantomData;

    use super::*;

    #[test]
    fn test_per_sample_block_adapter() {
        struct Echo<T>(PhantomData<T>);

        impl<T: Scalar> Echo<T> {
            pub fn new() -> Self {
                Self(PhantomData)
            }
        }

        impl<T: Scalar> DSPBlock<0, 1> for Echo<T> {
            type Sample = T;

            fn process_block(&mut self, inputs: &[[Self::Sample; 0]], outputs: &mut [[Self::Sample; 1]]) {
                let len = inputs.len();
                assert_eq!(len, outputs.len());

                for (i, [out]) in outputs.iter_mut().enumerate() {
                    *out = T::from_f64((i+1) as _);
                }
            }
        }

        let mut adaptor = PerSampleBlockAdapter::new_with_max_buffer_size(Echo::<f32>::new(), 4);
        assert_eq!(3, DSP::latency(&adaptor));

        let expected = [0.0, 0.0, 0.0, 1.0, 2.0, 3.0, 4.0, 1.0].map(|v| [v]);
        let mut actual = [[0.0]; 8];
        let input = [[]; 8];

        // Calling `process_block` but it's actually calling the impl for all `DSP` passing each sample through.
        adaptor.process_block(&input, &mut actual);

        assert_eq!(expected, actual);
    }
}