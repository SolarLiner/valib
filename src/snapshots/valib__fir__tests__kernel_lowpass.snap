---
source: src/fir.rs
expression: "&kernels::windowed_sinc(0.1, 0.1)"
---
-0.0
-0.00001
-0.00012
-0.00037
-0.00059
-0.0004
0.00065
0.00262
0.00482
0.00561
0.00301
-0.00413
-0.01465
-0.02424
-0.0261
-0.01329
0.01794
0.06531
0.11982
0.16794
0.19619
0.19619
0.16794
0.11982
0.06531
0.01794
-0.01329
-0.0261
-0.02424
-0.01465
-0.00413
0.00301
0.00561
0.00482
0.00262
0.00065
-0.0004
-0.00059
-0.00037
-0.00012
-0.00001
